### PR TITLE
APIS-4192 - Use stride auth

### PR DESCRIPTION
**Pull request by rose-garratt**

### CODE REVIEW

File: test/unit/uk/gov/hmrc/thirdpartyapplication/controllers/AuthorisationWrapperSpec.scala [hunk 1]
```diff
@@ -66,49 +60,49 @@ class AuthorisationWrapperSpec extends UnitSpec with MockitoSugar with WithFakeA
...
-      status(underTest.requiresRoleFor(APIGatekeeper, ROPC).async(BodyParsers.parse.json)(_ => Default.Ok(""))(ropcRequest)) shouldBe SC_OK
+    "accept the request when access type in the payload is ROPC and user is authenticated" in new Setup {
+      givenUserIsAuthenticated(underTest)
+      status(underTest.requiresRoleFor(ROPC).async(BodyParsers.parse.json)(_ => Default.Ok(""))(ropcRequest)) shouldBe SC_OK
     }
 
     "skip gatekeeper authentication for payload with STANDARD applications" in new Setup {
 
-      val result = await(underTest.requiresRoleFor(APIGatekeeper, PRIVILEGED).async(BodyParsers.parse.json)(_ =>
+      val result = await(underTest.requiresRoleFor(PRIVILEGED).async(BodyParsers.parse.json)(_ =>
```
Comment (FluentSynergyDW) [inline test/unit/uk/gov/hmrc/thirdpartyapplication/controllers/AuthorisationWrapperSpec.scala]: Says "STANDARD" in test name but uses "PRIVILEGED" in test code. Is that right?

Comment (FluentSynergyDW) [reply test/unit/uk/gov/hmrc/thirdpartyapplication/controllers/AuthorisationWrapperSpec.scala]: What does this test mean?

File: test/unit/uk/gov/hmrc/thirdpartyapplication/controllers/AuthorisationWrapperSpec.scala [hunk 2]
```diff
@@ -122,57 +116,57 @@ class AuthorisationWrapperSpec extends UnitSpec with MockitoSugar with WithFakeA
...
+      givenUserIsAuthenticated(underTest)
+      status(underTest.requiresRoleFor(applicationId, ROPC).async(_ => Default.Ok(""))(FakeRequest())) shouldBe SC_OK
     }
 
     "skip gatekeeper authentication for STANDARD applications" in new Setup {
 
       mockFetchApplicationToReturn(applicationId, Some(standardApplication))
 
-      val result = await(underTest.requiresRoleFor(applicationId, APIGatekeeper, PRIVILEGED).async(_ => Default.Ok(""))(FakeRequest()))
+      val result = await(underTest.requiresRoleFor(applicationId, PRIVILEGED).async(_ => Default.Ok(""))(FakeRequest()))
```
Comment (FluentSynergyDW) [inline test/unit/uk/gov/hmrc/thirdpartyapplication/controllers/AuthorisationWrapperSpec.scala]: Says "STANDARD" in test name but uses PRIVILEGED in code.

File: test/unit/uk/gov/hmrc/thirdpartyapplication/controllers/AuthorisationWrapperSpec.scala [hunk 3]
```diff
@@ -183,21 +177,21 @@ class AuthorisationWrapperSpec extends UnitSpec with MockitoSugar with WithFakeA
...
 
-      val result = await(underTest.requiresRole(APIGatekeeper).async(_ => Default.Ok(""))(FakeRequest()))
+      val result = await(underTest.requiresRole().async(_ => Default.Ok(""))(FakeRequest()))
 
-      status(result) shouldBe SC_UNAUTHORIZED
-      jsonBodyOf(result) shouldBe JsErrorResponse(UNAUTHORIZED, "Action requires authority: 'api:gatekeeper'")
+      status(result) shouldBe SC_FORBIDDEN
+      jsonBodyOf(result) shouldBe JsErrorResponse(FORBIDDEN, "Insufficient enrolments")
     }
   }
```
Comment (FluentSynergyDW) [inline test/unit/uk/gov/hmrc/thirdpartyapplication/controllers/AuthorisationWrapperSpec.scala]: I think this whole spec needs a rethink. It strikes me that maybe the behaviour being tested and the language being used no longer match the requirements.

File: test/unit/uk/gov/hmrc/thirdpartyapplication/controllers/GatekeeperControllerSpec.scala [hunk 4]
```diff
@@ -193,29 +198,27 @@ class GatekeeperControllerSpec extends UnitSpec with ScalaFutures with MockitoSu
...
-      val hcArgCaptor = ArgumentCaptor.forClass(classOf[HeaderCarrier])
-      when(mockAuthConnector.authorized(any[AuthRole])(hcArgCaptor.capture())).thenReturn(successful(true))
+      givenUserIsAuthenticated(underTest)
 
       when(mockGatekeeperService.rejectUplift(applicationId, rejectUpliftRequest)).thenReturn(UpliftRejected)
 
       val result = await(underTest.rejectUplift(applicationId)(testReq))
 
-      hcArgCaptor.getValue.authorization.get.value shouldBe "authorizationToken"
       status(result) shouldBe 204
```
Comment (FluentSynergyDW) [inline test/unit/uk/gov/hmrc/thirdpartyapplication/controllers/GatekeeperControllerSpec.scala]: Can we change this 204 to NO_CONTENT? And any other similar "magic number" occurences.

File: app/uk/gov/hmrc/thirdpartyapplication/controllers/AuthorisationWrapper.scala [hunk 5]
```diff
@@ -38,49 +40,51 @@ trait AuthorisationWrapper {
 
   val authConnector: AuthConnector
   val applicationService: ApplicationService
+  val authConfig: AuthConfig
 
-  def requiresRole(requiredRole: AuthRole): ActionBuilder[Request] = {
-    Action andThen AuthenticatedAction(requiredRole)
+  def requiresRole(): ActionBuilder[Request] = {
+    Action andThen AuthenticatedAction()
```
[the remark this answers is not in the document]
Comment (rose-garratt) [reply app/uk/gov/hmrc/thirdpartyapplication/controllers/AuthorisationWrapper.scala]: Potentially to rename, potentially to authenticated... TBD

File: app/uk/gov/hmrc/thirdpartyapplication/config/ConfigurationProviders.scala [hunk 6]
```diff
@@ -177,7 +177,10 @@ class AuthConfigProvider @Inject()(val runModeConfiguration: Configuration, envi
 
   override def get() = {
     val url = baseUrl("auth")
-    AuthConfig(url)
+    val userRole = getString("roles.user")
+    val superUserRole = getString("roles.super-user")
+    val adminRole = getString("roles.admin")
+    AuthConfig(url, userRole, superUserRole, adminRole)
```
[the remark this answers is not in the document]
Comment (rose-garratt) [reply app/uk/gov/hmrc/thirdpartyapplication/config/ConfigurationProviders.scala]: Confirmed with Nohmaan that admin role reqires extra permissions within gatekeeper as compared to super-user. This along with only one role allowed per user means we have to know about it.

File: app/uk/gov/hmrc/thirdpartyapplication/controllers/AuthorisationWrapper.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (rose-garratt) [reply app/uk/gov/hmrc/thirdpartyapplication/controllers/AuthorisationWrapper.scala]: Will rename to requiresAuthentication... also other similar functions.

File: app/uk/gov/hmrc/thirdpartyapplication/controllers/AuthorisationWrapper.scala [hunk 7]
```diff
@@ -38,49 +40,51 @@ trait AuthorisationWrapper {
...
-          Some(Unauthorized(JsErrorResponse(UNAUTHORIZED,
-            s"Action requires authority: '${authRole.scope}:${authRole.name}'")))
+  private abstract class AuthenticationFilter() extends ActionFilter[Request] {
+    def authenticate[A](request: Request[A]): Future[Option[Result]] = {
+      implicit val hc = HeaderCarrierConverter.fromHeadersAndSession(request.headers, None)
+      val hasAnyGatekeeperRole = Enrolment(authConfig.userRole) or Enrolment(authConfig.superUserRole) or Enrolment(authConfig.adminRole)
+      authConnector.authorise(hasAnyGatekeeperRole, EmptyRetrieval).map {
+        _ => None
+      } recover { case _ => Some(Forbidden(JsErrorResponse(FORBIDDEN,
+        s"Insufficient enrolments")))
```
[the remark this answers is not in the document]
Comment (rose-garratt) [reply app/uk/gov/hmrc/thirdpartyapplication/controllers/AuthorisationWrapper.scala]: The reason for failure in the authentication filter is always that the user hasn't got the authorisation (enrolment) to do something. I think this is appropriate.

File: conf/application.conf [hunk 8]
```diff
@@ -129,6 +129,12 @@ metrics {
 clientSecretLimit = 5
 upliftVerificationValidity = 90d
 
+roles {
+  super-user = "super-user-role"
+  user = "user-role"
+  admin = "admin-role"
+}
```
[the remark this answers is not in the document]
Comment (rose-garratt) [reply conf/application.conf]: not yet

File: test/unit/uk/gov/hmrc/thirdpartyapplication/controllers/ApplicationControllerSpec.scala [hunk 9]
```diff
@@ -171,28 +175,27 @@ class ApplicationControllerSpec extends UnitSpec with ScalaFutures with MockitoS
...
       val result = await(underTest.create()(request.withBody(Json.toJson(privilegedApplicationRequest))))
 
-      verifyErrorResult(result, SC_UNAUTHORIZED, ErrorCode.UNAUTHORIZED)
+      verifyErrorResult(result, SC_FORBIDDEN, ErrorCode.FORBIDDEN)
       verify(underTest.applicationService, never()).create(any[CreateApplicationRequest])(any[HeaderCarrier])
     }
 
-    "fail with a 401 (Unauthorized) for a valid ROPC application request when gatekeeper is not logged in" in new Setup {
-      when(underTest.authConnector.authorized(mockEq(APIGatekeeper))(any[HeaderCarrier])).thenReturn(false)
+    "fail with a 403 (Forbidden) for a valid ROPC application request when gatekeeper is not logged in" in new Setup {
```
[the remark this answers is not in the document]
Comment (rose-garratt) [reply test/unit/uk/gov/hmrc/thirdpartyapplication/controllers/ApplicationControllerSpec.scala]: Agreed, we'll change it back

File: test/unit/uk/gov/hmrc/thirdpartyapplication/controllers/AuthorisationWrapperSpec.scala (hunk 1 above)
Comment (rose-garratt) [reply test/unit/uk/gov/hmrc/thirdpartyapplication/controllers/AuthorisationWrapperSpec.scala]: It uses a standardRequest and requires authorisation if it were privileged, therefore it doesn't call need to make a call to auth. Have renamed for clarity, hopefully.

File: test/unit/uk/gov/hmrc/thirdpartyapplication/controllers/GatekeeperControllerSpec.scala (hunk 4 above)
Comment (rose-garratt) [reply test/unit/uk/gov/hmrc/thirdpartyapplication/controllers/GatekeeperControllerSpec.scala]: There are loads of magic numbers in this file. Wary of going down the rabbit hole of cleaning up untouched files.

**Comment by adampridmore** [comment]
This PR have been superseded by https://github.com/hmrc/third-party-application/pull/13. 
We'll delete it when PR-13 has been merged.

**Pull request closed without merging**